packages/playwright-core/src/server/socksClientCertificatesInterceptor.ts

```diff
@@ -84,51 +123,65 @@ class SocksProxyConnection {
-        internalTLS.end();
-        targetTLS.end();
-      };
-
-      internalTLS.on('end', () => closeBothSockets());
-      targetTLS.on('end', () => closeBothSockets());
-
-      internalTLS.on('error', () => closeBothSockets());
-      targetTLS.on('error', error => {
-        const responseBody = 'Playwright client-certificate error: ' + error.message;
-        internalTLS.end([
-          'HTTP/1.1 503 Internal Server Error',
-          'Content-Type: text/html; charset=utf-8',
-          'Content-Length: ' + Buffer.byteLength(responseBody),
-          '\r\n',
-          responseBody,
-        ].join('\r\n'));
-        closeBothSockets();
+    alpnCache.get(rewriteToLocalhostIfNeeded(this.host), this.port, serverRequestedAlpnProtocol => {
+      debugLogger.log('socks', `Proxy->Target ${this.host}:${this.port} chooses ALPN ${serverRequestedAlpnProtocol}`);
+      const dummyServer = tls.createServer({
+        key: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/key.pem')),
+        cert: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/cert.pem')),
+        ALPNProtocols: serverRequestedAlpnProtocol === 'h2' ? ['h2', 'http/1.1'] : ['http/1.1'],
       });
+      dummyServer.emit('connection', this.internal);
+      dummyServer.on('secureConnection', internalTLS => {
+        debugLogger.log('socks', `Browser->Proxy ${this.host}:${this.port} chooses ALPN ${internalTLS.alpnProtocol}`);
+        internalTLS.on('close', () => this.socksProxy._socksProxy.sendSocketEnd({ uid: this.uid }));
+        const tlsOptions: tls.ConnectionOptions = {
+          socket: this.target,
+          host: this.host,
+          port: this.port,
+          rejectUnauthorized: !this.socksProxy.ignoreHTTPSErrors,
+          ALPNProtocols: [internalTLS.alpnProtocol || 'http/1.1'],
+          ...clientCertificatesToTLSOptions(this.socksProxy.clientCertificates, `https://${this.host}:${this.port}/`),
+        };
+        if (!net.isIP(this.host))
+          tlsOptions.servername = this.host;
+        if (process.env.PWTEST_UNSUPPORTED_CUSTOM_CA && isUnderTest())
+          tlsOptions.ca = [fs.readFileSync(process.env.PWTEST_UNSUPPORTED_CUSTOM_CA)];
+        const targetTLS = tls.connect(tlsOptions);
+
+        internalTLS.pipe(targetTLS);
+        targetTLS.pipe(internalTLS);
+
+        // Handle close and errors
+        const closeBothSockets = () => {
+          internalTLS.end();
+          targetTLS.end();
+        };
+
+        internalTLS.on('end', () => closeBothSockets());
+        targetTLS.on('end', () => closeBothSockets());
+
+        internalTLS.on('error', () => closeBothSockets());
+        targetTLS.on('error', error => {
+          debugLogger.log('socks', `error when connecting to target: ${error.message}`);
+          if (internalTLS?.alpnProtocol === 'h2') {
+          // https://github.com/nodejs/node/issues/46152
+          // TODO: http2.performServerHandshake does not work here for some reason.
```

Comment:
As per offline discussion with dgozman, we might follow-up and spending more time on it.